Source/Core/Core/HW/SI/SI_DeviceAMBaseboard.cpp

```diff
@@ -0,0 +1,2375 @@
+  m_ic_card_data[0x28] = 0xFF;
+  m_ic_card_data[0x29] = 0xFF;
+
+  // Setup CARD
+  m_card_memory_size = 0;
+  m_card_is_inserted = false;
+
+  m_card_offset = 0;
+  m_card_command = 0;
+  m_card_clean = 0;
+
+  m_card_write_length = 0;
+  m_card_wrote = 0;
+
+  m_card_read_length = 0;
+  m_card_read = 0;
+
+  m_card_bit = 0;
+  m_card_shutter = true;  // Open
+  m_card_state_call_count = 0;
+
+  // Serial
+  m_wheel_init = 0;
+
+  m_motor_init = 0;
+  m_motor_force_y = 0;
+
+  m_fzdx_seatbelt = true;
+  m_fzdx_motion_stop = false;
+  m_fzdx_sensor_right = false;
+  m_fzdx_sensor_left = false;
+
+  m_rx_reply = 0xF0;
+
+  m_fzcc_seatbelt = true;
+  m_fzcc_sensor = false;
+  m_fzcc_emergency = false;
+  m_fzcc_service = false;
+
+  memset(m_motor_reply, 0, sizeof(m_motor_reply));
+}
+
+constexpr u32 SI_XFER_LENGTH_MASK = 0x7f;
+
+// Translate [0,1,2,...,126,127] to [128,1,2,...,126,127]
+constexpr s32 ConvertSILengthField(u32 field)
+{
+  return ((field - 1) & SI_XFER_LENGTH_MASK) + 1;
+}
+
+void CSIDevice_AMBaseboard::ICCardSendReply(ICCommand* iccommand, u8* buffer, u32* length)
+{
+  iccommand->status = Common::swap16(iccommand->status);
+
+  const u8 crc = CheckSumXOR(iccommand->data + 2, iccommand->pktlen - 1);
+
+  for (u32 i = 0; i < iccommand->pktlen + 1; ++i)
+  {
+    buffer[(*length)++] = iccommand->data[i];
+  }
+
+  buffer[(*length)++] = crc;
+}
+
+void CSIDevice_AMBaseboard::SwapBuffers(u8* buffer, u32* buffer_length)
+{
+  memcpy(m_last[1], buffer, 0x80);     // Save current buffer
+  memcpy(buffer, m_last[0], 0x80);     // Load previous buffer
+  memcpy(m_last[0], m_last[1], 0x80);  // Update history
+
+  m_lastptr[1] = *buffer_length;  // Swap lengths
+  *buffer_length = m_lastptr[0];
+  m_lastptr[0] = m_lastptr[1];
+}
+
+int CSIDevice_AMBaseboard::RunBuffer(u8* buffer, int request_length)
+{
+  const auto& serial_interface = m_system.GetSerialInterface();
+  u32 buffer_length = ConvertSILengthField(serial_interface.GetInLength());
+
+  // Debug logging
+  ISIDevice::RunBuffer(buffer, buffer_length);
+
+  u32 buffer_position = 0;
+  while (buffer_position < buffer_length)
+  {
+    BaseBoardCommand command = static_cast<BaseBoardCommand>(buffer[buffer_position]);
+    buffer_position++;
+
+    switch (command)
+    {
+    case BaseBoardCommand::GCAM_Reset:  // Returns ID and dip switches
+    {
+      const u32 id = Common::swap32(SI_AM_BASEBOARD | 0x100);
+      std::memcpy(buffer, &id, sizeof(id));
+      return sizeof(id);
+    }
+    break;
+    case BaseBoardCommand::GCAM_Command:
+    {
+      u32 checksum = 0;
+      for (u32 i = 0; i < buffer_length; ++i)
+        checksum += buffer[i];
+
+      std::array<u8, 0x80> data_out{};
+      u32 data_offset = 0;
+
+      static u32 dip_switch_1 = 0xFE;
+      static u32 dip_switch_0 = 0xFF;
+
+      data_out[data_offset++] = 1;
+      data_out[data_offset++] = 1;
+
+      u8* data_in = buffer + 2;
+      const u8* data_in_end = buffer + buffer[buffer_position] + 2;
```

Comment:
You aren't checking that `buffer_position` is smaller than `128`, and then you aren't checking that `buffer[buffer_position] + 2` is less than or equal to `128`.